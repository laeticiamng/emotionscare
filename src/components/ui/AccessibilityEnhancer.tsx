import React, { useEffect, useState } from 'react';
import { motion, AnimatePresence } from 'framer-motion';
import { Button } from '@/components/ui/button';
import { Badge } from '@/components/ui/badge';
import { Card, CardContent } from '@/components/ui/card';
import { 
  Eye, 
  Volume2, 
  Keyboard, 
  MousePointer, 
  Settings,
  Contrast,
  Type,
  ZoomIn,
  X,
  Check,
  AlertTriangle
} from 'lucide-react';
import { useAccessibility } from '@/components/common/AccessibilityProvider';
import {
  safeClassAdd,
  safeClassRemove,
  safeGetDocumentRoot
} from '@/lib/safe-helpers';

interface AccessibilitySettings {
  highContrast: boolean;
  reducedMotion: boolean;
  largeText: boolean;
  dyslexicFont: boolean;
  screenReader: boolean;
  keyboardNavigation: boolean;
  focusIndicators: boolean;
  colorBlindFriendly: boolean;
  announcements: boolean;
}

interface AccessibilityIssue {
  type: 'error' | 'warning' | 'info';
  message: string;
  element?: string;
  fix?: () => void;
}

export function AccessibilityEnhancer() {
  const [isOpen, setIsOpen] = useState(false);
  const [settings, setSettings] = useState<AccessibilitySettings>({
    highContrast: false,
    reducedMotion: false,
    largeText: false,
    dyslexicFont: false,
    screenReader: false,
    keyboardNavigation: true,
    focusIndicators: true,
    colorBlindFriendly: false,
    announcements: true
  });
  const [issues, setIssues] = useState<AccessibilityIssue[]>([]);
  const { announceToScreenReader } = useAccessibility();
  
  // Functions pour la compatibilité
  const announce = announceToScreenReader;
  const handleSkipLink = (id: string) => {
    if (typeof document === 'undefined') return;
    const element = document.getElementById(id);
    if (element) {
      element.focus();
      element.scrollIntoView({ behavior: 'smooth' });
    }
  };
  const generateId = () => `accessibility-${Math.random().toString(36).substr(2, 9)}`;

  useEffect(() => {
    // Load settings from localStorage
    if (typeof window === 'undefined') return;

    try {
      const savedSettings = window.localStorage.getItem('accessibility-settings');
      if (savedSettings) {
        setSettings(JSON.parse(savedSettings));
      }
    } catch (error) {
      console.warn('[AccessibilityEnhancer UI] Failed to read settings', error);
    }

    // Check for accessibility issues
    checkAccessibilityIssues();
  }, []);

  useEffect(() => {
    // Apply settings to document
    applyAccessibilitySettings(settings);
    // Save settings to localStorage
    try {
      if (typeof window !== 'undefined') {
        window.localStorage.setItem('accessibility-settings', JSON.stringify(settings));
      }
    } catch (error) {
      console.warn('[AccessibilityEnhancer UI] Failed to persist settings', error);
    }
  }, [settings]);

  const applyAccessibilitySettings = (newSettings: AccessibilitySettings) => {
    const root = safeGetDocumentRoot();
    const body = typeof document !== 'undefined' ? document.body : null;

    // High contrast
    if (newSettings.highContrast) {
      safeClassAdd(body, 'high-contrast');
    } else {
      safeClassRemove(body, 'high-contrast');
    }

    // Reduced motion
    if (newSettings.reducedMotion) {
      safeClassAdd(body, 'reduce-motion');
    } else {
      safeClassRemove(body, 'reduce-motion');
    }

    // Large text
    if (newSettings.largeText) {
      root.style.setProperty('--font-scale', '1.25');
    } else {
      root.style.setProperty('--font-scale', '1');
    }

    // Dyslexic font
    if (newSettings.dyslexicFont) {
      safeClassAdd(body, 'dyslexic-font');
    } else {
      safeClassRemove(body, 'dyslexic-font');
    }

    // Color blind friendly
    if (newSettings.colorBlindFriendly) {
      safeClassAdd(body, 'color-blind-friendly');
    } else {
      safeClassRemove(body, 'color-blind-friendly');
    }

    // Focus indicators
    if (newSettings.focusIndicators) {
      root.style.setProperty('--focus-ring-width', '3px');
    } else {
      root.style.setProperty('--focus-ring-width', '2px');
    }
  };

  const checkAccessibilityIssues = () => {
    if (typeof document === 'undefined') {
      setIssues([]);
      return;
    }
    const foundIssues: AccessibilityIssue[] = [];

    // Check for images without alt text
    const images = document.querySelectorAll('img:not([alt])');
    if (images.length > 0) {
      foundIssues.push({
        type: 'error',
        message: `${images.length} image(s) sans attribut alt détectée(s)`,
        element: 'Images',
        fix: () => {
          images.forEach((img, index) => {
            img.setAttribute('alt', `Image ${index + 1}`);
          });
          announce('Attributs alt ajoutés aux images manquantes');
        }
      });
    }

    // Check for buttons without accessible names
    const unnamedButtons = document.querySelectorAll('button:not([aria-label]):not([aria-labelledby])');
    const buttonIssues = Array.from(unnamedButtons).filter(btn => !btn.textContent?.trim());
    if (buttonIssues.length > 0) {
      foundIssues.push({
        type: 'warning',
        message: `${buttonIssues.length} bouton(s) sans nom accessible`,
        element: 'Boutons'
      });
    }

    // Check for form inputs without labels
    const unlabeledInputs = document.querySelectorAll('input:not([aria-label]):not([aria-labelledby])');
    const inputIssues = Array.from(unlabeledInputs).filter(input => {
      const id = input.getAttribute('id');
      return !id || !document.querySelector(`label[for="${id}"]`);
    });
    if (inputIssues.length > 0) {
      foundIssues.push({
        type: 'error',
        message: `${inputIssues.length} champ(s) de formulaire sans label`,
        element: 'Formulaires'
      });
    }

    // Check heading hierarchy
    const headings = Array.from(document.querySelectorAll('h1, h2, h3, h4, h5, h6'));
    const h1Count = headings.filter(h => h.tagName === 'H1').length;
    if (h1Count === 0) {
      foundIssues.push({
        type: 'warning',
        message: 'Aucun titre principal H1 trouvé',
        element: 'Titres'
      });
    } else if (h1Count > 1) {
      foundIssues.push({
        type: 'warning',
        message: `${h1Count} titres H1 trouvés (recommandation: 1 seul)`,
        element: 'Titres'
      });
    }

    setIssues(foundIssues);
  };

  const toggleSetting = (key: keyof AccessibilitySettings) => {
    setSettings(prev => {
      const newSettings = { ...prev, [key]: !prev[key] };
      
      // Announce the change
      const settingName = {
        highContrast: 'Contraste élevé',
        reducedMotion: 'Mouvement réduit',
        largeText: 'Texte agrandi',
        dyslexicFont: 'Police dyslexie',
        screenReader: 'Lecteur d\'écran',
        keyboardNavigation: 'Navigation clavier',
        focusIndicators: 'Indicateurs de focus',
        colorBlindFriendly: 'Mode daltonien',
        announcements: 'Annonces'
      }[key];

      announce(`${settingName} ${newSettings[key] ? 'activé' : 'désactivé'}`);
      
      return newSettings;
    });
  };

  const settingsGroups = [
    {
      title: 'Vision',
      icon: Eye,
      settings: [
        { key: 'highContrast' as const, label: 'Contraste élevé', description: 'Augmente le contraste pour une meilleure lisibilité' },
        { key: 'largeText' as const, label: 'Texte agrandi', description: 'Augmente la taille du texte de 25%' },
        { key: 'dyslexicFont' as const, label: 'Police dyslexie', description: 'Utilise une police adaptée à la dyslexie' },
        { key: 'colorBlindFriendly' as const, label: 'Mode daltonien', description: 'Palette de couleurs adaptée au daltonisme' }
      ]
    },
    {
      title: 'Mouvement',
      icon: MousePointer,
      settings: [
        { key: 'reducedMotion' as const, label: 'Mouvement réduit', description: 'Réduit les animations et transitions' }
      ]
    },
    {
      title: 'Navigation',
      icon: Keyboard,
      settings: [
        { key: 'keyboardNavigation' as const, label: 'Navigation clavier', description: 'Optimise la navigation au clavier' },
        { key: 'focusIndicators' as const, label: 'Indicateurs de focus', description: 'Améliore la visibilité du focus clavier' }
      ]
    },
    {
      title: 'Assistance',
      icon: Volume2,
      settings: [
        { key: 'screenReader' as const, label: 'Support lecteur d\'écran', description: 'Optimise pour les lecteurs d\'écran' },
        { key: 'announcements' as const, label: 'Annonces vocales', description: 'Annonce les changements importants' }
      ]
    }
  ];

  const criticalIssues = issues.filter(issue => issue.type === 'error');
  const warnings = issues.filter(issue => issue.type === 'warning');

  return (
    <>
      {/* Accessibility Skip Links */}
      <div className="skip-links">
        <a 
          href="#main-content" 
          className="skip-link"
          onClick={(e) => {
            e.preventDefault();
            handleSkipLink('main-content');
          }}
        >
          Aller au contenu principal
        </a>
        <a 
          href="#navigation" 
          className="skip-link"
          onClick={(e) => {
            e.preventDefault();
            handleSkipLink('navigation');
          }}
        >
          Aller à la navigation
        </a>
      </div>

      {/* Accessibility Button */}
      <Button
        onClick={() => setIsOpen(true)}
        className="fixed bottom-4 left-4 z-50 rounded-full p-3 shadow-lg"
        aria-label="Ouvrir les options d'accessibilité"
        title="Options d'accessibilité"
      >
        <Eye className="h-5 w-5" />
        {(criticalIssues.length > 0 || warnings.length > 0) && (
          <Badge className="absolute -top-1 -right-1 h-5 w-5 p-0 text-xs">
            {criticalIssues.length + warnings.length}
          </Badge>
        )}
      </Button>

      {/* Accessibility Panel */}
      <AnimatePresence mode="sync">
        {isOpen && (
          <>
            {/* Backdrop */}
            <motion.div
              initial={{ opacity: 0 }}
              animate={{ opacity: 1 }}
              exit={{ opacity: 0 }}
              className="fixed inset-0 bg-black/50 z-50"
              onClick={() => setIsOpen(false)}
            />

            {/* Panel */}
            <motion.div
              initial={{ x: -400, opacity: 0 }}
              animate={{ x: 0, opacity: 1 }}
              exit={{ x: -400, opacity: 0 }}
              transition={{ type: "spring", damping: 30, stiffness: 300 }}
              className="fixed left-0 top-0 bottom-0 w-96 bg-background border-r shadow-2xl z-50 overflow-y-auto"
              role="dialog"
              aria-label="Panneau d'accessibilité"
              aria-modal="true"
            >
              <div className="p-6 space-y-6">
                {/* Header */}
                <div className="flex items-center justify-between">
                  <h2 className="text-xl font-bold flex items-center gap-2">
                    <Eye className="h-6 w-6 text-primary" />
                    Accessibilité
                  </h2>
                  <Button
                    variant="ghost"
                    size="sm"
                    onClick={() => setIsOpen(false)}
                    aria-label="Fermer le panneau d'accessibilité"
                  >
                    <X className="h-4 w-4" />
                  </Button>
                </div>

                {/* Issues Summary */}
                {issues.length > 0 && (
                  <Card className={`${criticalIssues.length > 0 ? 'border-red-200 bg-red-50 dark:bg-red-950' : 'border-yellow-200 bg-yellow-50 dark:bg-yellow-950'}`}>
                    <CardContent className="p-4">
                      <div className="flex items-start gap-3">
                        <AlertTriangle className={`h-5 w-5 mt-0.5 ${criticalIssues.length > 0 ? 'text-red-500' : 'text-yellow-500'}`} />
                        <div className="flex-1">
                          <h3 className="font-semibold mb-2">
                            {criticalIssues.length > 0 ? 'Problèmes critiques détectés' : 'Améliorations suggérées'}
                          </h3>
                          <div className="space-y-2">
                            {issues.slice(0, 3).map((issue, index) => (
                              <div key={index} className="flex items-start justify-between gap-2">
                                <span className="text-sm">{issue.message}</span>
                                {issue.fix && (
                                  <Button size="sm" variant="outline" onClick={issue.fix}>
                                    Corriger
                                  </Button>
                                )}
                              </div>
                            ))}
                          </div>
                          {issues.length > 3 && (
                            <p className="text-xs text-muted-foreground mt-2">
                              +{issues.length - 3} autre(s) problème(s)
                            </p>
                          )}
                        </div>
                      </div>
                    </CardContent>
                  </Card>
                )}

                {/* Quick Actions */}
                <div className="space-y-3">
                  <h3 className="font-semibold">Actions rapides</h3>
                  <div className="grid grid-cols-2 gap-2">
                    <Button
                      variant="outline"
                      size="sm"
                      onClick={() => toggleSetting('highContrast')}
                      className="justify-start gap-2"
                    >
                      <Contrast className="h-4 w-4" />
                      Contraste
                    </Button>
                    <Button
                      variant="outline"
                      size="sm"
                      onClick={() => toggleSetting('largeText')}
                      className="justify-start gap-2"
                    >
                      <Type className="h-4 w-4" />
                      Texte+
                    </Button>
                    <Button
                      variant="outline"
                      size="sm"
                      onClick={() => toggleSetting('reducedMotion')}
                      className="justify-start gap-2"
                    >
                      <MousePointer className="h-4 w-4" />
                      Mouvement-
                    </Button>
                    <Button
                      variant="outline"
                      size="sm"
                      onClick={checkAccessibilityIssues}
                      className="justify-start gap-2"
                    >
                      <Settings className="h-4 w-4" />
                      Analyser
                    </Button>
                  </div>
                </div>

                {/* Settings Groups */}
                {settingsGroups.map((group) => {
                  const Icon = group.icon;
                  return (
                    <div key={group.title} className="space-y-3">
                      <h3 className="font-semibold flex items-center gap-2">
                        <Icon className="h-5 w-5 text-primary" />
                        {group.title}
                      </h3>
                      <div className="space-y-3">
                        {group.settings.map((setting) => (
                          <div key={setting.key} className="flex items-start justify-between gap-3">
                            <div className="flex-1">
                              <label 
                                htmlFor={setting.key}
                                className="text-sm font-medium cursor-pointer"
                              >
                                {setting.label}
                              </label>
                              <p className="text-xs text-muted-foreground">
                                {setting.description}
                              </p>
                            </div>
                            <Button
                              id={setting.key}
                              variant={settings[setting.key] ? "default" : "outline"}
                              size="sm"
                              onClick={() => toggleSetting(setting.key)}
                              aria-pressed={settings[setting.key]}
                              className="shrink-0"
                            >
                              {settings[setting.key] ? <Check className="h-3 w-3" /> : <X className="h-3 w-3" />}
                            </Button>
                          </div>
                        ))}
                      </div>
                    </div>
                  );
                })}

                {/* Info */}
                <div className="text-xs text-muted-foreground p-4 bg-muted rounded-lg">
                  <p className="mb-2">
                    <strong>Raccourcis clavier :</strong>
                  </p>
                  <ul className="space-y-1">
                    <li>• Tab : Navigation entre éléments</li>
                    <li>• Entrée/Espace : Activer un élément</li>
                    <li>• Échap : Fermer les dialogues</li>
                    <li>• Alt + A : Ouvrir ce panneau</li>
                  </ul>
                </div>
              </div>
            </motion.div>
          </>
        )}
      </AnimatePresence>

      {/* Keyboard shortcut listener */}
      {typeof window !== 'undefined' && (
        <KeyboardShortcuts onOpenAccessibility={() => setIsOpen(true)} />
      )}
    </>
  );
}

// Keyboard shortcuts component
function KeyboardShortcuts({ onOpenAccessibility }: { onOpenAccessibility: () => void }) {
  useEffect(() => {
    const handleKeyDown = (event: KeyboardEvent) => {
      // Alt + A to open accessibility panel
      if (event.altKey && event.key === 'a') {
        event.preventDefault();
        onOpenAccessibility();
      }
    };

    document.addEventListener('keydown', handleKeyDown);
    return () => document.removeEventListener('keydown', handleKeyDown);
  }, [onOpenAccessibility]);

  return null;
}