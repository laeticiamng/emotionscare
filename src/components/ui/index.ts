export * from './button';
export * from './card';
export * from './form';
export * from './input';
export * from './label';
export * from './skeleton';
export * from './toast';
export * from './toaster';
export * from './loading-spinner';
export * from './loading-animation';
export * from './enhanced-error-boundary';
export * from './loading-illustration';
export * from './badge';
export * from './sheet';
export * from './tabs';
export * from './separator';
export * from './switch';
export * from './card';
export * from './form';
