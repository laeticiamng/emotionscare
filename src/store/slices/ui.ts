import type { StateCreator } from 'zustand';
import type { AppState } from '../appStore';

export type Theme = 'light' | 'dark' | 'system';
export type ModuleKey = 'music' | 'emotion' | 'journal' | 'coach';

export interface UIState {
  theme: Theme;
  sidebarCollapsed: boolean;
  activeModule: ModuleKey | null;
}

export interface UIActions {
  setTheme: (theme: Theme) => void;
  toggleSidebar: () => void;
  setActiveModule: (module: ModuleKey | null) => void;
}

export type UISlice = UIState & UIActions;

export const createUIInitialState = (): UIState => ({
  theme: 'system',
  sidebarCollapsed: false,
  activeModule: null,
});

export const createUISlice: StateCreator<AppState, [], [], UISlice> = (set, get) => ({
  ...createUIInitialState(),
  setTheme: (theme) => set({ theme }),
  toggleSidebar: () => set({ sidebarCollapsed: !get().sidebarCollapsed }),
  setActiveModule: (module) => set({ activeModule: module }),
});
