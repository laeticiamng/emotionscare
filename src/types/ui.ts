
export interface Toast {
  id: string;
  title?: string;
  description?: string;
  action?: React.ReactNode;
  variant?: 'default' | 'destructive' | 'success';
  duration?: number;
}

export interface ToastAction {
  altText: string;
  onClick: () => void;
  children?: React.ReactNode;
}

export interface ThemeSwitcherProps {
  size?: 'default' | 'sm' | 'lg';
  showLabel?: boolean;
}
