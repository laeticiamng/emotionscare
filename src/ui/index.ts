export * from './Alert';
export * from './Badge';
export * from './Button';
export * from './Card';
export * from './ChartA11yCaption';
export * from './EmptyState';
export * from './Form';
export * from './Input';
export * from './Modal';
export * from './Navbar';
export * from './Spinner';
export * from './a11y/FocusRing';
export * from './a11y/SkipLink';
export * from './a11y/VisuallyHidden';
