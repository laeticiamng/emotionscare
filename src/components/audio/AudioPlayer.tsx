import React, { useState, useRef, useEffect } from 'react';
import { Card, CardContent } from '@/components/ui/card';
import { Button } from '@/components/ui/button';
import { Slider } from '@/components/ui/slider';
import { Play, Pause, SkipBack, SkipForward, Volume2, VolumeX, Repeat, Shuffle } from 'lucide-react';
import { Badge } from '@/components/ui/badge';

interface AudioTrack {
  id: string;
  title: string;
  artist: string;
  duration: number;
  url: string;
  artwork?: string;
}

interface AudioPlayerProps {
  tracks: AudioTrack[];
  currentTrackIndex?: number;
  onTrackChange?: (index: number) => void;
  autoPlay?: boolean;
  showPlaylist?: boolean;
}

const AudioPlayer: React.FC<AudioPlayerProps> = ({
  tracks,
  currentTrackIndex = 0,
  onTrackChange,
  autoPlay = false,
  showPlaylist = true
}) => {
  const [isPlaying, setIsPlaying] = useState(false);
  const [currentTime, setCurrentTime] = useState(0);
  const [volume, setVolume] = useState(0.7);
  const [isMuted, setIsMuted] = useState(false);
  const [isRepeat, setIsRepeat] = useState(false);
  const [isShuffle, setIsShuffle] = useState(false);
  const [currentIndex, setCurrentIndex] = useState(currentTrackIndex);
  
  const audioRef = useRef<HTMLAudioElement>(null);
  const progressRef = useRef<HTMLDivElement>(null);

  const currentTrack = tracks[currentIndex];

  useEffect(() => {
    if (audioRef.current) {
      audioRef.current.volume = isMuted ? 0 : volume;
    }
  }, [volume, isMuted]);

  useEffect(() => {
    if (audioRef.current && currentTrack) {
      audioRef.current.src = currentTrack.url;
      if (autoPlay) {
        audioRef.current.play();
        setIsPlaying(true);
      }
    }
  }, [currentIndex, currentTrack, autoPlay]);

  const togglePlay = () => {
    if (audioRef.current) {
      if (isPlaying) {
        audioRef.current.pause();
      } else {
        audioRef.current.play();
      }
      setIsPlaying(!isPlaying);
    }
  };

  const handlePrevious = () => {
    const newIndex = currentIndex > 0 ? currentIndex - 1 : tracks.length - 1;
    setCurrentIndex(newIndex);
    onTrackChange?.(newIndex);
  };

  const handleNext = () => {
    let newIndex;
    if (isShuffle) {
      newIndex = Math.floor(Math.random() * tracks.length);
    } else {
      newIndex = currentIndex < tracks.length - 1 ? currentIndex + 1 : 0;
    }
    setCurrentIndex(newIndex);
    onTrackChange?.(newIndex);
  };

  const handleTimeUpdate = () => {
    if (audioRef.current) {
      setCurrentTime(audioRef.current.currentTime);
    }
  };

  const handleSeek = (value: number[]) => {
    if (audioRef.current && currentTrack) {
      const newTime = (value[0] / 100) * currentTrack.duration;
      audioRef.current.currentTime = newTime;
      setCurrentTime(newTime);
    }
  };

  const handleVolumeChange = (value: number[]) => {
    setVolume(value[0] / 100);
    setIsMuted(false);
  };

  const toggleMute = () => {
    setIsMuted(!isMuted);
  };

  const formatTime = (seconds: number) => {
    const mins = Math.floor(seconds / 60);
    const secs = Math.floor(seconds % 60);
    return `${mins}:${secs.toString().padStart(2, '0')}`;
  };

  const handleEnded = () => {
    if (isRepeat) {
      audioRef.current?.play();
    } else {
      handleNext();
    }
  };

  if (!currentTrack) return null;

  return (
    <div className="space-y-4">
      <audio
        ref={audioRef}
        onTimeUpdate={handleTimeUpdate}
        onEnded={handleEnded}
        onLoadedMetadata={() => {
          if (audioRef.current) {
            setCurrentTime(0);
          }
        }}
      />
      
      <Card>
        <CardContent className="p-6">
          <div className="flex items-center gap-4 mb-4">
            {currentTrack.artwork && (
              <img
                src={currentTrack.artwork}
                alt={currentTrack.title}
                className="w-16 h-16 rounded-lg object-cover"
              />
            )}
            <div className="flex-1">
              <h3 className="font-semibold text-lg">{currentTrack.title}</h3>
              <p className="text-muted-foreground">{currentTrack.artist}</p>
            </div>
            <Badge variant="secondary">
              {formatTime(currentTime)} / {formatTime(currentTrack.duration)}
            </Badge>
          </div>
          
          {/* Progress Bar */}
          <div className="mb-4">
            <Slider
              value={[currentTrack.duration > 0 ? (currentTime / currentTrack.duration) * 100 : 0]}
              onValueChange={handleSeek}
              max={100}
              step={0.1}
              className="w-full"
            />
          </div>
          
          {/* Controls */}
          <div className="flex items-center justify-center gap-2 mb-4">
            <Button
              variant="ghost"
              size="icon"
              onClick={() => setIsShuffle(!isShuffle)}
              className={isShuffle ? 'text-primary' : ''}
            >
              <Shuffle className="h-4 w-4" />
            </Button>
            
            <Button variant="ghost" size="icon" onClick={handlePrevious}>
              <SkipBack className="h-4 w-4" />
            </Button>
            
            <Button size="icon" onClick={togglePlay}>
              {isPlaying ? <Pause className="h-5 w-5" /> : <Play className="h-5 w-5" />}
            </Button>
            
            <Button variant="ghost" size="icon" onClick={handleNext}>
              <SkipForward className="h-4 w-4" />
            </Button>
            
            <Button
              variant="ghost"
              size="icon"
              onClick={() => setIsRepeat(!isRepeat)}
              className={isRepeat ? 'text-primary' : ''}
            >
              <Repeat className="h-4 w-4" />
            </Button>
          </div>
          
          {/* Volume Control */}
          <div className="flex items-center gap-2">
            <Button variant="ghost" size="icon" onClick={toggleMute}>
              {isMuted ? <VolumeX className="h-4 w-4" /> : <Volume2 className="h-4 w-4" />}
            </Button>
            <Slider
              value={[isMuted ? 0 : volume * 100]}
              onValueChange={handleVolumeChange}
              max={100}
              step={1}
              className="w-24"
            />
          </div>
        </CardContent>
      </Card>
      
      {/* Playlist */}
      {showPlaylist && tracks.length > 1 && (
        <Card>
          <CardContent className="p-4">
            <h4 className="font-semibold mb-3">Playlist ({tracks.length} titres)</h4>
            <div className="space-y-2 max-h-60 overflow-y-auto">
              {tracks.map((track, index) => (
                <div
                  key={track.id}
                  className={`flex items-center gap-3 p-2 rounded-lg cursor-pointer transition-colors ${
                    index === currentIndex ? 'bg-primary/10 text-primary' : 'hover:bg-muted'
                  }`}
                  onClick={() => {
                    setCurrentIndex(index);
                    onTrackChange?.(index);
                  }}
                >
                  <div className="w-8 h-8 rounded bg-muted flex items-center justify-center text-xs">
                    {index + 1}
                  </div>
                  <div className="flex-1 min-w-0">
                    <p className="font-medium truncate">{track.title}</p>
                    <p className="text-sm text-muted-foreground truncate">{track.artist}</p>
                  </div>
                  <span className="text-xs text-muted-foreground">
                    {formatTime(track.duration)}
                  </span>
                </div>
              ))}
            </div>
          </CardContent>
        </Card>
      )}
    </div>
  );
};

export default AudioPlayer;
