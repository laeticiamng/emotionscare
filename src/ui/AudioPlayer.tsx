"use client";
import React from "react";
import { useSound } from "@/ui/hooks/useSound";
import { clamp01 } from "@/lib/audio/utils";

type Props = {
  src: string;
  title?: string;
  loop?: boolean;
  defaultVolume?: number; // 0..1
  haptics?: boolean;
};

export function AudioPlayer({ src, title, loop, defaultVolume = 0.8, haptics = false }: Props) {
  const s = useSound(src, { loop, volume: defaultVolume });
  const [playing, setPlaying] = React.useState(false);
  const [volume, setVolume] = React.useState(defaultVolume);

  React.useEffect(() => { s.setVolume?.(volume); }, [volume]);

  async function toggle() {
    try {
      if (!playing) {
        await s.play?.();
        setPlaying(true);
        const reduced = typeof window !== "undefined" && window.matchMedia("(prefers-reduced-motion: reduce)").matches;
        if (haptics && !reduced && "vibrate" in navigator) navigator.vibrate?.(10);
      } else {
        s.pause?.();
        setPlaying(false);
      }
    } catch {}
  }

  return (
    <div aria-label={title ?? "Lecteur audio"} style={{ display: "grid", gap: 8 }}>
      <div style={{ display: "flex", gap: 8, alignItems: "center" }}>
        <button onClick={toggle} aria-pressed={playing} data-ui="primary-cta">
          {playing ? "Pause" : "Lecture"}
        </button>
        {title && <strong>{title}</strong>}
      </div>
      <label style={{ display: "flex", gap: 8, alignItems: "center" }}>
        Volume
        <input
          type="range" min={0} max={1} step={0.01} value={volume}
          onChange={(e) => setVolume(clamp01(parseFloat(e.target.value)))}
          data-ui="volume-slider"
        />
      </label>
    </div>
  );
}
