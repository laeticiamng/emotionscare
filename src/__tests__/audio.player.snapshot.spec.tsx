import { describe, it, expect, vi } from "vitest";
import { fireEvent, render, screen } from "@testing-library/react";

import { AudioPlayer } from "@/ui/AudioPlayer";

describe("AudioPlayer", () => {
  it("renders, toggles favorite state and exposes volume slider", async () => {
    const toggle = vi.fn();
    const { container } = render(
      <AudioPlayer
        src="/audio/lofi-120.mp3"
        trackId="test-track"
        title="Test"
        favorite={{ active: false, onToggle: toggle }}
      />,
    );

    expect(screen.getByLabelText("Volume")).toBeInTheDocument();
    const favorite = screen.getByRole("button", { name: /Ajouter aux favoris/i });
    fireEvent.click(favorite);
    expect(toggle).toHaveBeenCalledTimes(1);
    expect(container).toMatchSnapshot();
  });

  it("shows resume button when resume prop provided", () => {
    render(
      <AudioPlayer
        src="/audio/lofi-120.mp3"
        trackId="test-track"
        title="Test"
        resume={{ position: 42, onResume: vi.fn() }}
      />,
    );

    expect(screen.getByRole("button", { name: /Reprendre/ })).toHaveTextContent("0:42");
    expect(screen.getByText(/Dernière écoute sauvegardée/)).toBeInTheDocument();
  });
});
