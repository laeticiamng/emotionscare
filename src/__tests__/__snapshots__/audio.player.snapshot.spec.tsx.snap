// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AudioPlayer > renders and exposes volume slider 1`] = `
<div>
  <div
    aria-label="Test"
    style="display: grid; gap: 8px;"
  >
    <div
      style="display: flex; gap: 8px; align-items: center;"
    >
      <button
        aria-pressed="false"
        data-ui="primary-cta"
      >
        Lecture
      </button>
      <strong>
        Test
      </strong>
    </div>
    <label
      style="display: flex; gap: 8px; align-items: center;"
    >
      Volume
      <input
        data-ui="volume-slider"
        max="1"
        min="0"
        step="0.01"
        type="range"
        value="0.8"
      />
    </label>
  </div>
</div>
`;
